torch/_higher_order_ops/scan.py

```diff
@@ -181,56 +157,35 @@ def _check_new_carry_match_init(leaves_init, carry_leaves):
     combine_fn = functools.partial(
         wrap_combine_fn_flat,
         combine_fn=combine_fn,
-        spec_init=spec_init,
-        spec_xs=spec_xs,
-        num_init_leaves=len(leaves_init),
-        num_inp_leaves=len(leaves_xs),
+        spec_init=init_spec,
+        spec_xs=xs_spec,
+        num_init_leaves=len(flat_init),
+        num_inp_leaves=len(flat_xs),
     )
 
     def run_flattened_scan(combine_fn, leaves_init, leaves_xs, dim, reverse):
         return scan_op(
-            combine_fn, leaves_init, leaves_xs, dim, reverse, additional_inputs=[]
+            combine_fn, leaves_init, leaves_xs, dim=dim, reverse=reverse, additional_inputs=[]
         )
 
-    if not torch._dynamo.is_compiling():
-        from torch._dynamo.backends.debugging import (
-            make_eager_backend_with_torch_function_mode,
+    carry, result = _maybe_compile_and_run_fn(run_flattened_scan, combine_fn, flat_init, flat_xs, dim, reverse)
+    
+    # TODO: Maybe this check is already too late and should be done earlier?
+    # Check the tree specs of the carry and the init
+    if not init_spec.__eq__(pytree.tree_structure(carry)):
+        raise RuntimeError(
+            "The pytree of the carry needs to match the pytree of the init"
         )
 
-        with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
-            with _temp_remove_metadata_torch_function_mode() as metadata_mode:
-                if metadata_mode:
-                    backend = make_eager_backend_with_torch_function_mode(metadata_mode)
-                else:
-                    backend = "eager"
-                result = torch.compile(
-                    run_flattened_scan, backend=backend, fullgraph=True
-                )(
-                    combine_fn,
-                    leaves_init,
-                    leaves_xs,
-                    dim=dim,
-                    reverse=reverse,
-                )
-    else:
-        result = run_flattened_scan(combine_fn, leaves_init, leaves_xs, dim, reverse)
-
-    result_carry, result_flat = _extract_carry_and_out(
-        result,
-        len(leaves_init),
-    )
-
-    return pytree.tree_unflatten(result_carry, spec_init), pytree.tree_unflatten(
-        result_flat, tree_out
-    )
+    return carry, result
 
 
 class ScanOp(HigherOrderOperator):
     def __init__(self):
         super().__init__("scan")
 
     def __call__(self, combine_fn, init, xs, dim, reverse, additional_inputs):
-        assert isinstance(additional_inputs, list), "additional_inputs must be a list."
+        assert isinstance(additional_inputs, tuple), "additional_inputs must be a list."
```

Comment:
maybe change the comment

Reply:
This was actually a change by mistake. I kept the additional_inputs to be a list